hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/hudi/functional/TestSparkSqlWithCustomKeyGenerator.scala

```diff
@@ -365,10 +373,63 @@ class TestSparkSqlWithCustomKeyGenerator extends HoodieSparkSqlTestBase {
         // INSERT INTO should succeed now
         testFirstRoundInserts(tableName, TS_FORMATTER_FUNC, customPartitionFunc)
       }
+
+      // Validate ts field is still of type int in the table
+      validateTsFieldSchema(tablePath)
     }
     }
   }
 
+  test("Test query with custom key generator") {
+    withTempDir { tmp => {
+      val tableName = generateTableName
+      val tablePath = tmp.getCanonicalPath + "/" + tableName
+      val writePartitionFields = "ts:timestamp"
+      val dateFormat = "yyyy/MM/dd"
+      val tsGenFunc = (ts: Integer) => TS_FORMATTER_FUNC_WITH_FORMAT.apply(ts, dateFormat)
+      val customPartitionFunc = (ts: Integer, _: String) => tsGenFunc.apply(ts)
+      val keyGenConfigs = TS_KEY_GEN_CONFIGS + ("hoodie.keygen.timebased.output.dateformat" -> dateFormat)
+
+      prepareTableWithKeyGenerator(
+        tableName, tablePath, "MERGE_ON_READ",
+        CUSTOM_KEY_GEN_CLASS_NAME, writePartitionFields, keyGenConfigs)
+
+      createTableWithSql(tableName, tablePath,
+        s"hoodie.datasource.write.partitionpath.field = '$writePartitionFields', "
+          + keyGenConfigs.map(e => e._1 + " = '" + e._2 + "'").mkString(", "))
+
+      // INSERT INTO should fail due to conflict between write and table config of partition path fields
+      val sourceTableName = tableName + "_source"
+      prepareParquetSource(sourceTableName, Seq("(7, 'a7', 1399.0, 1706800227, 'cat1')"))
+      // INSERT INTO should succeed now
+      testFirstRoundInserts(tableName, tsGenFunc, customPartitionFunc)
+      assertEquals(7, spark.sql(
+        s"""
+           | SELECT * from $tableName
+           | """.stripMargin).count())
+      val incrementalDF = spark.read.format("hudi").
+        option("hoodie.datasource.query.type", "incremental").
+        option("hoodie.datasource.read.begin.instanttime", 0).
+        load(tablePath)
+      incrementalDF.createOrReplaceTempView("tbl_incremental")
+      assertEquals(7, spark.sql(
+        s"""
+           | SELECT * from tbl_incremental
+           | """.stripMargin).count())
+
+      // Validate ts field is still of type int in the table
+      validateTsFieldSchema(tablePath)
+    }
+    }
+  }
+
+  private def validateTsFieldSchema(tablePath: String): Unit = {
```

Comment:
we can make it more reusable, by taking field name and expected type in the argument.

Reply:
Addressed